// react-instantsearch/es/src/core/createInstantSearchServer.js
import _isEmpty from 'lodash/isEmpty';

var _createClass = function () { function defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; }();

var _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; };

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _possibleConstructorReturn(self, call) { if (!self) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return call && (typeof call === "object" || typeof call === "function") ? call : self; }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function, not " + typeof superClass); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, enumerable: false, writable: true, configurable: true } }); if (superClass) Object.setPrototypeOf ? Object.setPrototypeOf(subClass, superClass) : subClass.__proto__ = superClass; }

function _toConsumableArray(arr) { if (Array.isArray(arr)) { for (var i = 0, arr2 = Array(arr.length); i < arr.length; i++) { arr2[i] = arr[i]; } return arr2; } else { return Array.from(arr); } }

import PropTypes from 'prop-types';
import React, { Component } from 'react';
var _name$version$descrip = {
  name: 'react-instantsearch',
  version: '4.4.1',
  description: '\u26A1 Lightning-fast search for React and React Native apps, by Algolia',
  keywords: ['algolia', 'components', 'fast', 'instantsearch', 'react', 'react-native', 'search'],
  homepage: 'https://community.algolia.com/react-instantsearch',
  license: 'MIT',
  author: {
    name: 'Algolia, Inc.',
    url: 'https://www.algolia.com'
  },
  main: 'index.js',
  module: 'es/index.js',
  repository: {
    type: 'git',
    url: 'https://github.com/algolia/react-instantsearch'
  },
  scripts: {
    build: './scripts/build.sh',
    'build-and-publish': './scripts/build-and-publish.sh'
  },
  dependencies: {
    algoliasearch: '^3.24.0',
    'algoliasearch-helper': '^2.21.0',
    classnames: '^2.2.5',
    lodash: '^4.17.4',
    'prop-types': '^15.5.10'
  },
  devDependencies: {
    enzyme: '3.3.0',
    'enzyme-adapter-react-16': '1.1.1',
    react: '16.2.0',
    'react-dom': '16.2.0',
    'react-native': '0.51.0',
    'react-test-renderer': '16.2.0'
  }
},
    version = _name$version$descrip.version;

import algoliasearchHelper, { SearchResults, SearchParameters } from 'algoliasearch-helper';
import ReactDom from 'react-dom/server';
import { getIndex, hasMultipleIndex } from './indexUtils';

import cis from './createInstantSearch';
import highlightTags from './highlightTags.js';

var createInstantSearch = function createInstantSearch(algoliasearch) {
  var InstantSearch = cis(algoliasearch, {
    Root: 'div',
    props: { className: 'ais-InstantSearch__root' }
  });

  var searchParameters = [];
  var client = void 0;
  var indexName = '';

  var onSearchParameters = function onSearchParameters(getSearchParameters, context, props, searchState) {
    var index = getIndex(context);
    searchParameters.push({
      getSearchParameters: getSearchParameters,
      context: context,
      props: props,
      searchState: searchState,
      index: index
    });
  };

  var findResultsState = function findResultsState(App, props) {
    searchParameters = [];
    ReactDom.renderToString(React.createElement(App, props));
    var sharedSearchParameters = searchParameters.filter(function (searchParameter) {
      return !hasMultipleIndex(searchParameter.context) && (searchParameter.props.indexName === indexName || !searchParameter.props.indexName);
    }).reduce(function (acc, searchParameter) {
      return searchParameter.getSearchParameters.call({ context: searchParameter.context }, acc, searchParameter.props, searchParameter.searchState);
    }, new SearchParameters(_extends({ index: indexName }, highlightTags)));

    var mergedSearchParameters = searchParameters.filter(function (searchParameter) {
      return hasMultipleIndex(searchParameter.context);
    }).reduce(function (acc, searchParameter) {
      var index = getIndex(searchParameter.context);
      var sp = searchParameter.getSearchParameters.call({ context: searchParameter.context }, acc[index] ? acc[index] : sharedSearchParameters, searchParameter.props, searchParameter.searchState);
      acc[index] = sp;
      return acc;
    }, {});

    if (_isEmpty(mergedSearchParameters)) {
      var helper = algoliasearchHelper(client, sharedSearchParameters.index);
      return helper.searchOnce(sharedSearchParameters);
    } else {
      var _helper = algoliasearchHelper(client, sharedSearchParameters.index);
      var search = [];
      if (mergedSearchParameters[indexName]) {
        search.push(_helper.searchOnce(_extends({}, sharedSearchParameters, mergedSearchParameters[sharedSearchParameters.index], {
          index: mergedSearchParameters[sharedSearchParameters.index].index
        })));
        delete mergedSearchParameters[indexName];
      } else {
        search.push(_helper.searchOnce(sharedSearchParameters));
      }
      search.push.apply(search, _toConsumableArray(Object.keys(mergedSearchParameters).map(function (key) {
        var derivedHelper = algoliasearchHelper(client, mergedSearchParameters[key].index);
        return derivedHelper.searchOnce(mergedSearchParameters[key]);
      })));

      return Promise.all(search);
    }
  };

  var decorateResults = function decorateResults(results) {
    if (!results) {
      return undefined;
    }
    return Array.isArray(results) ? results.reduce(function (acc, result) {
      acc[result.state.index] = new SearchResults(new SearchParameters(result.state), result._originalResponse.results);
      return acc;
    }, []) : new SearchResults(new SearchParameters(results.state), results._originalResponse.results);
  };

  var CreateInstantSearchServer = function (_Component) {
    _inherits(CreateInstantSearchServer, _Component);

    function CreateInstantSearchServer(props) {
      _classCallCheck(this, CreateInstantSearchServer);

      var _this = _possibleConstructorReturn(this, (CreateInstantSearchServer.__proto__ || Object.getPrototypeOf(CreateInstantSearchServer)).call(this));

      client = props.algoliaClient || algoliasearch(props.appId, props.apiKey);
      client.addAlgoliaAgent('react-instantsearch ' + version);
      indexName = props.indexName;
      return _this;
    }

    _createClass(CreateInstantSearchServer, [{
      key: 'render',
      value: function render() {
        var resultsState = decorateResults(this.props.resultsState);
        return React.createElement(InstantSearch, _extends({}, this.props, {
          resultsState: resultsState,
          onSearchParameters: onSearchParameters
        }));
      }
    }]);

    return CreateInstantSearchServer;
  }(Component);

  CreateInstantSearchServer.propTypes = {
    algoliaClient: PropTypes.object,
    appId: PropTypes.string,
    apiKey: PropTypes.string,
    indexName: PropTypes.string.isRequired,
    resultsState: PropTypes.oneOfType([PropTypes.object, PropTypes.array])
  };


  return { InstantSearch: CreateInstantSearchServer, findResultsState: findResultsState };
};

export { createInstantSearch };
